### PR TITLE
More Coordinate tests for variously lazy/real points and bounds.

### PR DESCRIPTION
**Pull request by pp-mo**
Following the changes in #2527

### CODE REVIEW

File: lib/iris/tests/unit/coords/test_AuxCoord.py [hunk 1]
```diff
@@ -31,22 +31,72 @@
...
+import dask.array as da
+import mock
 import numpy as np
+import unittest
 
-from iris._lazy_data import as_lazy_data, is_lazy_data
 from iris.coords import AuxCoord
 
 
+_DEBUG = True
```
Comment (dkillick) [inline lib/iris/tests/unit/coords/test_AuxCoord.py]: Assumedly all of these additions will not make it into the final code?

File: lib/iris/tests/unit/coords/test_DimCoord.py [hunk 2]
```diff
@@ -23,11 +31,75 @@
...
 
+import dask.array as da
+import mock
 import numpy as np
+import unittest
 
 from iris.coords import DimCoord
 
 
+_DEBUG = True
```
Comment (dkillick) [inline lib/iris/tests/unit/coords/test_DimCoord.py]: Again.

File: lib/iris/tests/unit/coords/test_DimCoord.py [hunk 3]
```diff
@@ -23,11 +31,75 @@
...
+
+
+_AUXCOORD_LAZY_API = hasattr(DimCoord, 'has_lazy_points')
```
Comment (dkillick) [inline lib/iris/tests/unit/coords/test_DimCoord.py]: I guess this should be called "DIMCOORD_LAZY_API"?

File: lib/iris/tests/unit/coords/test_AuxCoord.py (hunk 1 above)
Comment (pp-mo) [inline lib/iris/tests/unit/coords/test_AuxCoord.py]: Assuredly not !

File: lib/iris/tests/unit/coords/test_DimCoord.py (hunk 3 above)
Comment (pp-mo) [inline lib/iris/tests/unit/coords/test_DimCoord.py]: Yes

**Comment by pp-mo** [comment]
Updates: I ***think*** all these tests are now correct,
but some re-checking will be wise...

There are outstanding errors: 
I think they are real -- see [here](https://github.com/SciTools/iris/pull/2527/files#r115871042)

File: lib/iris/tests/unit/coords/test_AuxCoord.py [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+        coord = AuxCoord(self.pts_real, bounds=self.bds_real)
+        result = coord.lazy_bounds()
+        self.assertIsInstance(result, da.Array)
+        self.assertArrayEqual(result.dtype, self.bds_real.dtype)
+        self.assertArrayEqual(result.compute(), self.bds_real)
+
+    def test_lazy_core(self):
+        coord = AuxCoord(self.pts_real, bounds=self.bds_lazy)
+        result = coord.lazy_bounds()
+        self.assertIs(result, self.bds_lazy)
```
Comment (dkillick) [inline lib/iris/tests/unit/coords/test_AuxCoord.py]: Time for a controversial question: should you include a test for no bounds here too? Evidently, if the coord has no bounds then `lazy_bounds` will be `None`. I think there should be a test for that here, but if we do that then we may need to add equivalent tests elsewhere in these test modules...

**Review by DPeterK — COMMENTED** [review]
More comments!

**Comment by pp-mo** [comment]
I want to retarget this at upstream/dask, but you can't do that.
So this now becomes / continues as : https://github.com/SciTools/iris/pull/2547

File: lib/iris/tests/unit/coords/test_DimCoord.py [hunk 5]
```diff
@@ -0,0 +1,613 @@
+# (C) British Crown Copyright 2013 - 2017, Met Office
```
Comment (pp-mo) [inline lib/iris/tests/unit/coords/test_DimCoord.py]: Yes !
Done !!

File: lib/iris/tests/unit/coords/test_DimCoord.py [hunk 6]
```diff
@@ -0,0 +1,620 @@
...
+        condition=not _DIMCOORD_LAZY_API,
+        reason=('Test(s) only applicable to the enhanced DimCoord API '
+                '(lazy support).'))
+
+    return skip(fn)
+
+
+# Utility functions to get underlying points/bounds data, which work with old
+# coords as well as new.
+def _core_points(coord):
```
Comment (pp-mo) [inline lib/iris/tests/unit/coords/test_DimCoord.py]: All that "backwards compatibility" was only intended as a temporary measure while I didn't have the new coords code to test against.
So I am now removing ***all of it***  : 
 * skip_lazy_xxxcoords, _
 * _DEBUG, _xxxCOORD_LAZY_API
 * _coord_points/bounds()
I am doing this anyway,

Comment (pp-mo) [inline lib/iris/tests/unit/coords/test_DimCoord.py]: I will consider moving common code into `iris.tests.unit.coords`, but it will undoubtedly make them less readable.
